net.rocketpool/image/src/app.ts

```diff
@@ -1,163 +1,262 @@
+  const rethStartBlock = 13325304;
+
+  // Init
+  const providerUrl = env.ETH_NODE_PROVIDER_URL;
+
+  console.log(`ETH node provider URL: ${providerUrl}`);
+
+  const provider = getProvider(providerUrl)
+  const rethContract = RethAbi__factory.connect(rethAddress, provider)
+
+  // Block range
+  const lastSeenBlock = process.env.ODF_ETAG ? Number(process.env.ODF_ETAG) : 0;
+
+  console.log(`Last seen block: ${lastSeenBlock}`);
+
+  const ethHeadBlock = await provider.getBlockNumber();
+
+  console.log(`Current chain head block: ${ethHeadBlock}`);
+
+  const startBlock = Math.max(lastSeenBlock + 1, rethStartBlock);
+  const endBlock = ethHeadBlock;
+  const {
+    etag,
+    hasMore,
+  } = await odfReadRows(rethContract, startBlock, endBlock, env.BLOCK_BATCH_SIZE, env.ODF_BATCH_SIZE);
+
+  // Notify `kamu-cli` regarding new state
+  fs.writeFileSync(env.ODF_NEW_ETAG_PATH, etag.toString());
+
+  if (hasMore) {
+    fs.writeFileSync(env.ODF_NEW_HAS_MORE_DATA_PATH, "");
+  }
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+async function odfReadRows(
+  rethContract: RethAbi,
+  startBlock: number,
+  endBlock: number,
+  blockBatchSize: number,
+  odfBatchSize: number,
+): Promise<OdfReadRowsResult> {
+  let outputRowsCount = 0;
+  let lastOutputBlockNumber = 0;
+  let hasBatchAlreadyCollected = false;
+
+  const allBlockRangeHasProcessed = await readEventsFromETH(
+    rethContract,
     startBlock,
     endBlock,
-    endBlock - startBlock
+    function (eventRow) {
+      // In a usual situation (not web3), we could write the following condition:
+      // ```
+      // if (outputRowsCount >= odfBatchSize) {
+      //   return false;
+      // }
+      // ```
+      // But in our case, we cannot act so naively -- there can be multiple events in one block:
+      // ```
+      // ┌─────────────┐
+      // │ E1 E2 E3 E4 │
+      // └─────────────┘
+      //  Block №1
+      // ```
+      //
+      // This gives us an interesting edge case if the batch size goes through the middle of the block:
+      // ```
+      //  Batch №1  ┆ Batch №2
+      // ┌───────── ┆ ───┐  ┌─────────────┐
+      // │ E1 E2 E3 ┆ E4 │  │ E5 E6 E7 E8 │
+      // └───────── ┆ ───┘  └─────────────┘
+      //  Block №1  ┆        Block №2
+      // ```
+      //
+      // Since, our next read iteration relies on a range of blocks, this situation creates ambiguity for us:
+      // should we consider this block as processed or not?
+      //
+      // For this reason, we read additional events to consider this block as fully processed:
+      // ```
+      //  Batch №1         ┆  Batch №2
+      // ┌─────────   ───┐ ┆ ┌─────────────┐
+      // │ E1 E2 E3 + E4 │ ┆ │ E5 E6 E7 E8 │
+      // └─────────   ───┘ ┆ └─────────────┘
+      //  Block №1         ┆  Block №2
+      // ```
+      const isEventFromSameBlock = lastOutputBlockNumber === eventRow.blockNumber;
```

Comment:
This seems overcomplicated.

`ODF_BATCH_SIZE` is meant as a recommendation, not a hard restriction. So it's OK for a streaming source to return more records than specified. 

_// Some sources are simply non-resumable and will have to return all records in one go._

Given this, we could simply:
- iterate over block ranges
- read a batch of events
- write them to STDOUT
- terminate if we reached the `ODF_BATCH_SIZE`

Reply:
I understand that this extra code brings us closer to the recommended `ODF_BATCH_SIZE`, but since this container is intended as an example and not some prod component I would prefer to keep it as simple as possible. 

We want to show people how easy it is to implement the custom source, so extra complexity is better avoided. This was also the reason I used JS initially instead of TS - to make container look maximally simple.

Reply:
That totally makes sense for me, no problems -- simplified/done